### PR TITLE
Feat: vela dry-run render results should be affected by override policy and deploy workflowstep

### LINKED ISSUE #4410: [Feature] vela dry-run render results should be affected by override policy and deploy workflowstep
**Is your feature request related to a problem? Please describe.**

Now `vela dry-run` didn't take policy and workflow into consideration. But it will not show the real results especially for override policy.

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by cezhang**
Feat: vela dry-run render results should be affected by override policy and deploy workflowstep

Signed-off-by: cezhang <c1zhang.dev@gmail.com>

### Description of your changes

Fixes #4410

I have:

- [X] Read and followed KubeVela's [contribution process](https://github.com/kubevela/kubevela/blob/master/contribute/create-pull-request.md).
- [ ] [Related Docs](https://github.com/kubevela/kubevela.io) updated properly. In a new feature or configuration option, an update to the documentation is necessary. 
- [X] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### Special notes for your reviewer

Example: 
 Input application
```
apiVersion: core.oam.dev/v1beta1
kind: Application
metadata:
  name: first-vela-app
spec:
  components:
    - name: express-server
      type: webservice
      properties:
        image: oamdev/hello-world
        ports:
         - port: 8000
           expose: true
      traits:
        - type: scaler
          properties:
            replicas: 1
  policies:
    - name: target-default
      type: topology
      properties:
        # The cluster with name local is installed the KubeVela.
        clusters: ["local"]
        namespace: "default"
    - name: target-prod
      type: topology
      properties:
        clusters: ["local"]
        # This namespace must be created before deploying.
        namespace: "prod"
    - name: deploy-ha
      type: override
      properties:
        components:
          - type: webservice
            traits:
              - type: scaler
                properties:
                  replicas: 3
  workflow:
    steps:
      - name: deploy2default
        type: deploy
        properties:
          policies: ["target-default"]
      - name: manual-approval
        type: suspend
      - name: deploy2prod
        type: deploy
        properties:
          policies: ["target-prod", "deploy-ha"]
```
Result from dryrun command considering override policy:
```
---
# Application(first-vela-app) -- Component(express-server) 
---

apiVersion: apps/v1
kind: Deployment
metadata:
  annotations: {}
  labels:
    app.oam.dev/appRevision: ""
    app.oam.dev/component: express-server
    app.oam.dev/name: first-vela-app
    app.oam.dev/namespace: default
    app.oam.dev/resourceType: WORKLOAD
    workload.oam.dev/type: webservice
  name: express-server
  namespace: default
spec:
  replicas: 1
  selector:
    matchLabels:
      app.oam.dev/component: express-server
  template:
    metadata:
      labels:
        app.oam.dev/component: express-server
        app.oam.dev/name: first-vela-app
    spec:
      containers:
      - image: oamdev/hello-world
        name: express-server
        ports:
        - containerPort: 8000
          name: port-8000
          protocol: TCP

---
## From the auxiliary workload 
apiVersion: v1
kind: Service
metadata:
  annotations: {}
  labels:
    app.oam.dev/appRevision: ""
    app.oam.dev/component: express-server
    app.oam.dev/name: first-vela-app
    app.oam.dev/namespace: default
    app.oam.dev/resourceType: TRAIT
    trait.oam.dev/resource: webserviceExpose
    trait.oam.dev/type: AuxiliaryWorkload
  name: express-server
  namespace: default
spec:
  ports:
  - name: port-8000
    port: 8000
    targetPort: 8000
  selector:
    app.oam.dev/component: express-server
  type: ClusterIP

---

===
# The above components override by Policy(deploy-ha) below 
===

---
# Application(first-vela-app) -- Component(express-server) 
---

apiVersion: apps/v1
kind: Deployment
metadata:
  annotations: {}
  labels:
    app.oam.dev/appRevision: ""
    app.oam.dev/component: express-server
    app.oam.dev/name: first-vela-app
    app.oam.dev/namespace: default
    app.oam.dev/resourceType: WORKLOAD
    workload.oam.dev/type: webservice
  name: express-server
  namespace: default
spec:
  replicas: 3
  selector:
    matchLabels:
      app.oam.dev/component: express-server
  template:
    metadata:
      labels:
        app.oam.dev/component: express-server
        app.oam.dev/name: first-vela-app
    spec:
      containers:
      - image: oamdev/hello-world
        name: express-server
        ports:
        - containerPort: 8000
          name: port-8000
          protocol: TCP

---
## From the auxiliary workload 
apiVersion: v1
kind: Service
metadata:
  annotations: {}
  labels:
    app.oam.dev/appRevision: ""
    app.oam.dev/component: express-server
    app.oam.dev/name: first-vela-app
    app.oam.dev/namespace: default
    app.oam.dev/resourceType: TRAIT
    trait.oam.dev/resource: webserviceExpose
    trait.oam.dev/type: AuxiliaryWorkload
  name: express-server
  namespace: default
spec:
  ports:
  - name: port-8000
    port: 8000
    targetPort: 8000
  selector:
    app.oam.dev/component: express-server
  type: ClusterIP

---
```

### CODE REVIEW

File: pkg/appfile/dryrun/dryrun.go [hunk 1]
```diff
@@ -206,3 +210,33 @@ func (d *Option) PrintDryRun(buff *bytes.Buffer, appName string, comps []*types.
 	}
 	return nil
 }
+
+// ExecuteDryRunWithOverridePolicy is similar to ExecuteDryRun function, but patch components with override policy
+func (d *Option) ExecuteDryRunWithOverridePolicy(ctx context.Context, application *v1beta1.Application) (map[string][]*types.ComponentManifest, error) {
```
Comment (Somefive) [inline pkg/appfile/dryrun/dryrun.go:215]: Is this function aware of deploy step?

**Review by Somefive — COMMENTED** [review]
Seems you are making one copy of dryrun result for each override policy. There are some problems for that.

1. Override policy can be used jointly. Therefore, it is possible to use multiple override policies concurrently in one deploy step. This scenario is not taken into consideration AFAIS.
2. Declared override policy might not be used in workflow which means the dryrun result will not be actually deployed possibly.
3. Override policy can be referenced from the standalone override policy outside the application, it depends on how workflow refers to it.

Generally, I think the PR is a good attempt to tackle the difficult dynamic dryrun problem. But I recommend to consider the workflow deploy more carefully. Otherwise, this feature will cause confusions for users (since they will get different results between dryrun and real runs).

File: pkg/appfile/dryrun/diff_test.go [hunk 2]
```diff
@@ -158,16 +154,16 @@ var _ = Describe("Test Live-Diff", func() {
...
+		//	bs, err := ioutil.ReadFile("./testdata/" + filename)
+		//	Expect(err).Should(Succeed())
+		//	un := &unstructured.Unstructured{}
+		//	Expect(yaml.Unmarshal(bs, un)).Should(Succeed())
+		//	un.SetNamespace(ns)
+		//	Expect(k8sClient.Create(context.Background(), un)).Should(Succeed())
+		//}
 		ctx := context.Background()
-		Expect(k8sClient.Create(ctx, &corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: "vela-system"}})).Should(Succeed())
+		//Expect(k8sClient.Create(ctx, &corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: "vela-system"}})).Should(Succeed())
```
[the remark this answers is not in the document]
Comment (cezhang) [reply pkg/appfile/dryrun/diff_test.go:168]: already move to suite_test BeforeSuite function, my testcase needs this setting as well. Cannot guarantee which testcase  invoked first,  if both testcase have this setting,  error show.

File: pkg/appfile/dryrun/dryrun.go (hunk 1 above)
Comment (cezhang) [reply pkg/appfile/dryrun/dryrun.go:215]: if not using offline, yes

**Comment by cezhang** [comment]
> 

For the 3rd point, could u share a example?

**Comment by Somefive** [comment]
> > 
> 
> For the 3rd point, could u share a example?

https://kubevela.net/docs/end-user/components/ref-objects#distribute-reference-objects-with-different-configuration

File: pkg/appfile/dryrun/diff_test.go [hunk 3]
```diff
@@ -158,16 +156,16 @@ var _ = Describe("Test Live-Diff", func() {
...
 		}
-		applyFile := func(filename string, ns string) {
-			bs, err := ioutil.ReadFile("./testdata/" + filename)
-			Expect(err).Should(Succeed())
-			un := &unstructured.Unstructured{}
-			Expect(yaml.Unmarshal(bs, un)).Should(Succeed())
-			un.SetNamespace(ns)
-			Expect(k8sClient.Create(context.Background(), un)).Should(Succeed())
-		}
+		//applyFile := func(filename string, ns string) {
```
Comment (StevenLeiZhang) [inline pkg/appfile/dryrun/diff_test.go:159]: delete the useless codes

File: pkg/appfile/dryrun/dryrun.go [hunk 4]
```diff
@@ -206,3 +210,33 @@ func (d *Option) PrintDryRun(buff *bytes.Buffer, appName string, comps []*types.
...
+
+	compsForEechOverridePolicy := make(map[string][]*types.ComponentManifest)
```
Comment (StevenLeiZhang) [inline pkg/appfile/dryrun/dryrun.go:217]: Is "Eech" one typo? I guess you want "Each".

File: pkg/appfile/dryrun/dryrun.go [hunk 5]
```diff
@@ -206,3 +210,33 @@ func (d *Option) PrintDryRun(buff *bytes.Buffer, appName string, comps []*types.
...
+	for _, policy := range application.Spec.Policies {
+		if policy.Type == v1alpha1.OverridePolicyType {
```
Comment (StevenLeiZhang) [inline pkg/appfile/dryrun/dryrun.go:219]:
```
if policy.Type != v1alpha1.OverridePolicyType { continue }
/your main logic here/

```
Then you get less indent.

File: pkg/appfile/dryrun/dryrun.go (hunk 4 above)
Comment (cezhang) [reply pkg/appfile/dryrun/dryrun.go:217]: I am re-implementing this feature, btw, thanks for the head-ups

**Comment by cezhang** [comment]
@Somefive @wonderflow  please help take a quick review on the new impl

**Comment by wonderflow** [comment]
@cezhang What's the current behavior? Can you provide some examples that aligns with the design: https://github.com/kubevela/kubevela/discussions/5184#discussioncomment-4385727

**Comment by Somefive** [comment]
Great impl. Several suggestions for the change:
1. Let's reuse the `-f` parameter and let it accepts multiple file inputs, like `-f a.yaml,b.yaml,c.yaml` and `-f a.yaml -f b.yaml`, instead of adding the `-p` parameter. The PolicyFiles are not necessarily be treated differently. 
2. [Optional] Report error if multiple applications are found or no application found. Report error when resources other than Application/Policy/Workflow are included. Report error when multiple Workflows are provided. Report warnings if the external policy is not used in deploy workflow step.
3. [Optional] Support workflow file input.
4. [Optional] Support `--merge` to make external policy take effect when no workflow specified.

2~4 are recommended but we can leave their implementations for the future.

File: references/cli/dryrun_test.go [hunk 6]
```diff
@@ -17,125 +17,223 @@
...
+		opt := DryRunCmdOptions{ApplicationFiles: []string{"test-data/dry-run/testing-dry-run-5.yaml", "test-data/dry-run/testing-wf.yaml", "test-data/dry-run/testing-policy.yaml"}, OfflineMode: false, MergeOrphanFiles: true}
+		buff, err := DryRunApplication(&opt, c, "")
 		Expect(err).Should(BeNil())
+		Expect(buff.String()).Should(ContainSubstring("# Application(testing-app with topology deploy-somewhere)"))
+		Expect(buff.String()).Should(ContainSubstring("name: testing-dryrun"))
+		Expect(buff.String()).Should(ContainSubstring("kind: Deployment"))
+	})
 
-		dryRunOpt := dryrun.NewDryRunOption(k8sClient, cfg, dm, pd, nil, false)
+	It("Testing dry-run with orphan policy", func() {
```
Comment (Somefive) [inline references/cli/dryrun_test.go:200]: orphan -> standalone?

File: references/cli/dryrun.go [hunk 7]
```diff
@@ -241,3 +248,140 @@ func readApplicationFromFile(filename string) (*corev1beta1.Application, error)
...
+
+	for _, filename := range cmdOption.ApplicationFiles {
+		fileContent, err := utils.ReadRemoteOrLocalPath(filename, true)
+		if err != nil {
+			return nil, err
+		}
+
+		fileType := filepath.Ext(filename)
+		switch fileType {
+		case ".yaml", ".yml":
```
Comment (Somefive) [inline references/cli/dryrun.go:266]: only one case for fileType?

File: references/cli/dryrun.go [hunk 8]
```diff
@@ -241,3 +248,140 @@ func readApplicationFromFile(filename string) (*corev1beta1.Application, error)
...
+			// only support one object in one yaml file
+			fileContent, err = yaml.YAMLToJSON(fileContent)
```
Comment (Somefive) [inline references/cli/dryrun.go:268]: So we only support one object in one yaml file?

File: references/cli/dryrun.go [hunk 9]
```diff
@@ -45,9 +52,10 @@ import (
 // DryRunCmdOptions contains dry-run cmd options
 type DryRunCmdOptions struct {
 	cmdutil.IOStreams
-	ApplicationFile string
-	DefinitionFile  string
-	OfflineMode     bool
+	ApplicationFiles []string
+	DefinitionFile   string
+	OfflineMode      bool
+	MergeOrphanFiles bool
```
Comment (Somefive) [inline references/cli/dryrun.go:58]: MergeOrphanFiles -> MergeStandaloneFiles?

File: pkg/appfile/dryrun/diff_test.go [hunk 10]
```diff
@@ -190,7 +188,7 @@ var _ = Describe("Test Live-Diff", func() {
 		Expect(runDiff()).Should(ContainSubstring("\"myworker\" not found"))
 		applyFile("td-myingress.yaml", "vela-system")
 		applyFile("cd-myworker.yaml", "vela-system")
-		applyFile("wd-deploy.yaml", "vela-system")
+		//applyFile("wd-deploy.yaml", "vela-system")
```
Comment (Somefive) [inline pkg/appfile/dryrun/diff_test.go:191]: Let's just remove these two lines.

**Review by Somefive — COMMENTED** [review]
Generally LGTM after minor fixes, great job!

File: references/cli/dryrun.go (hunk 8 above)
Comment (cezhang) [reply references/cli/dryrun.go:268]: yes, only handle one object per file for now(yaml format, actually).  Plan to  add more support from up coming improvement

File: references/cli/dryrun.go [hunk 11]
```diff
@@ -88,9 +96,10 @@ You can also specify a remote url for app:
 		},
 	}
 
-	cmd.Flags().StringVarP(&o.ApplicationFile, "file", "f", "./app.yaml", "application file name")
+	cmd.Flags().StringSliceVarP(&o.ApplicationFiles, "files", "f", []string{"app.yaml"}, "application related file names")
```
Comment (wonderflow) [inline references/cli/dryrun.go:99]:
```suggestion
	cmd.Flags().StringSliceVarP(&o.ApplicationFiles, "file", "f", []string{"app.yaml"}, "application related file names")
```

We should keep the compatiblity, and naming it `file` with `--file filea --file fileb` is also reasonable

File: pkg/appfile/dryrun/diff_test.go [hunk 12]
```diff
@@ -167,7 +165,7 @@ var _ = Describe("Test Live-Diff", func() {
 			Expect(k8sClient.Create(context.Background(), un)).Should(Succeed())
 		}
 		ctx := context.Background()
-		Expect(k8sClient.Create(ctx, &corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: "vela-system"}})).Should(Succeed())
+		//Expect(k8sClient.Create(ctx, &corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: "vela-system"}})).Should(Succeed())
```
Comment (wonderflow) [inline pkg/appfile/dryrun/diff_test.go:168]: yes, please remove these lines to avoid confusion.

File: references/cli/dryrun.go [hunk 13]
```diff
@@ -45,9 +52,10 @@ import (
...
-	ApplicationFile string
-	DefinitionFile  string
-	OfflineMode     bool
+	ApplicationFiles []string
+	DefinitionFile   string
+	OfflineMode      bool
+	MergeOrphanFiles bool
 }
 
 // NewDryRunCommand creates `dry-run` command
```
Comment (wonderflow) [inline references/cli/dryrun.go:61]: let's add some explanation about the working mechanism.

1. Add some examples commands.
2. Dry Run with policy and workflow will only take `override/topology` policies and `deploy` workflow step into considerations. Other workflow step will be ignored.

File: references/cli/dryrun.go (hunk 8 above)
Comment (Somefive) [reply references/cli/dryrun.go:286]: Great, I strongly suggest supporting multiple files here. Although it is not necessary to be contained in this PR. And the multiple file loading could be moved into kubevela/pkg and be reused across various projects.

**Review by Somefive — APPROVED** [review]
LGTM, great job!

**Review by wonderflow — APPROVED** [review]
LGTM, please fix the CI

**Pull request merged**